README.rst
=========================
HTTP plugins for buildout
=========================

HTTP Basic-Authentication
=========================

With this extension it is possible to define password protected
package directories without specifying the password and user in the
url.

Let's take the example protected location, ``http://www.example.com/dist``

First we would need to add the extension and the find link for our
protected location::

    [buildout]
    find-links = http://www.example.com/dist
    extensions = lovely.buildouthttp

Then create the ``.httpauth`` password file, this file contains all
authentication information. The ``.httpauth`` file can be placed in the root of
the current buildout or in the ``~/.buildout`` directory. Each row consists of
``realm, uri, username, password``.

Here is an example of the ``.httpauth`` file::

    Example com realm, http://www.example.com, username, secret

It is also possible to leave the secret away. Then you will be prompted for the
secret whenever buildout is run::

    Example com realm, http://www.example.com, username

Note that basic auth also works with any recipe using
zc.buildout.download (e.g. hexagonit.recipe.download) because this
extension also overwrites the url opener of zc.buildout.

Github Private Downloads
========================

Private downloads on http://github.com/ require authorization to download.
The previous token-based authentication system based on the v2 API (see
http://github.com/blog/170-token-authentication) is no longer supported by
GitHub as of June 1 2012; You must now request a v3 API token and use that
instead.

Requesting a new API token can be done in one line using ``curl`` (please
substitute your own github username and password):

    curl -s -X POST -d '{"scopes": ["repo"], "note": "my API token"}' \
        https://${user}:${pass}@api.github.com/authorizations | grep token

Now set the value of github.token to the hash returned from the command above:

    git config --global github.accesstoken ${token}

Note that the v3 API does not require your github username to work, and can
be removed from your configuration if you wish.

For details on managing authorization GitHub's OAuth tokens, see the API
documentation: http://developer.github.com/v3/oauth/#oauth-authorizations-api

URL to download a tag or branch::

    https://api.github.com/repos/<gituser>/<repos>/tarball/master

URL to downlad a "download"::

    https://github.com/downloads/<gituser>/<repos>/<name>

As some eggs on PyPi also use public Github download URLs you may want to
whitelist the repos that authentication is required for as Github will
return a 401 error code even for public repositories if the wrong auth
details are provided.
To do this just list each repo in the format `<gituser>/<repos>` one per
line in the buildout config `github-repos`::

    [buildout]
    extensions = lovely.buildouthttp
    github-repos = lovelysystems/lovely.buildouthttp
                   bitly/asyncmongo


Development Setup
==============================

Run bootstrap::

    /path/to/python/python2.7 bootstrap.py

And run buildout::

    ./bin/buildout -N

Testing
_______

To run the tests use::

    ./bin/test

To run the tests with multiple python interpreters use tox::

    ./bin/tox

Credits
=======

Thanks to Tarek Ziade, Kevin Williams and Wesley Mason for bugfixes and extensions.


CHANGES.txt
===============================
Changes for lovely.buildouthttp
===============================

unreleased
==========

2015/01/14 0.6.1:
=================

 - added support for zc.buildout versions >= 2.0.0 (monkey patch of urlopener
   has changed)

2014/12/12 0.6.0:
=================

 - added python3 compatibility

2013/08/06 0.5.0:
=================

 - fix compatibility issues for setuptools 0.7 and later.

2013/06/25 0.4.8:
=================

 - the User-Agent header isn't dropped anymore because the Github API requires
   it.

 - prompt for password if it's not in the .httpauth file

2012/10/08 0.4.7:
=================

- also check the whitelist for github downloads (not just find-links)

- more testing

2012/09/21 0.4.6:
=================

- allow optional whitelist for github repositories (see README)

2012/06/21 0.4.5:
=================

- detect all github.com domain names

2012/06/20 0.4.4:
=================

- adapt github authentication to api v3 (see README.txt for more information)

2012/03/20 0.4.3:
=================

- fixed another problem related with the changed github authentication.

2012/03/19 0.4.2:
=================

- fixed github authentication
   According to Petros Amiridis (Github Staff), 'POST authentication is no
   longer supported.

2011/03/01 0.4.1:
=================

- exclude empty lines form .httpauth, this fixes issue 3.

2010/12/29 0.4.0:
=================

- Allow the inclusion of the .httpauth file from within the buildout itself.
  This adds to the realms set in ``~/.buildout/.httpauth``

2010/10/11 0.3.7:
=================

- bugfix: make git credentials work without any basic http auth
  definitions

2010/10/11 0.3.6:
=================

- added missing CHANGES.txt file, this fixes the installation

2010/10/09 0.3.5:
=================

- the github download credentials are now also added for
  zc.buildout.download based recipies. this allows to use github
  private download urls with any download recipe, such as
  hexagonit.recipe.download

- moved to github http://github.com/lovelysystems/lovely.buildouthttp

2010/09/28 0.3.4:
=================

- fixed behavour with python 2.6+. python 2.6 suddenly counts ``failed`` requests
  but each request without auth counts as one, that makes buildout fail on
  every 5th download in a single run.
  I found this on windows, but should fail on linux too.

2010/08/28 0.3.3:
=================

- fixed typo on git command, introduced in 0.3.2

2010/08/28 0.3.2:
=================

- silently ignore failing git credential lookups

2010/05/28 0.3.1:
=================

- python2.4 comptibility fix (thanks to Godefroid Chapelle)

2010/05/27 0.3.0:
=================

- added support for zc.buildout.download, this allows various download
  recipes to work with this credential extension. Requires at least
  zc.buildout version 1.5.0b2

2010/03/24 0.3.0a1:
===================

- added github authentication, which now allows to download from
  private github repositories

2009/09/22 0.2.3:
=================

- added tests for the installer

- fixed error when url can't be retrieved:
  AttributeError: 'NoneType' object has no attribute 'url'
  [naro]

2008/03/22 0.2.2:
=================

- added try/except around the retry section
  protection against weird URLs

2007/12/08 0.2.1:
=================

- avoid an error when the realm is wrong or the file on bad realms and a few
  other controls. Thanks to Tarek Ziade.

2007/08/06 0.2.0a1:
===================

- don't overwrite handlers in case a handler already exists. (e.g. in paramiko)
  thanks to Aaron Lehmann

2007/07/23 0.1.0a3:
===================

- added information to setup.py, moved to cheeseshop

2007/06/21 0.1.0a2:
===================

- Log download failures

- Fixed license header

- Log connections on debug.

2007/06/14 0.1.0a1:
===================

- initial release


src/lovely/buildouthttp/tests.py
import doctest
import unittest
from doctest import DocFileSuite, DocTestSuite
import tempfile
import shutil


def setUp(test):
    test.globs['tmp'] = tempfile.mkdtemp()


def tearDown(test):
    shutil.rmtree(test.globs['tmp'])


def test_suite():
    uSuites = (
        DocFileSuite(
            'install.txt',
            setUp=setUp,
            tearDown=tearDown,
            optionflags=doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS,
            ),
        DocTestSuite(
            'lovely.buildouthttp.buildouthttp',
            setUp=setUp, tearDown=tearDown,
            optionflags=doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS,
            ),
        )
    return unittest.TestSuite(uSuites)


trunk/src/lovely/buildouthttp/tests.py
import doctest, unittest
from zope.testing.doctest import DocTestSuite
import tempfile
import shutil

def setUp(test):
    test.globs['tmp'] = tempfile.mkdtemp()

def tearDown(test):
    shutil.rmtree(test.globs['tmp'])


def test_suite():
    uSuites = (
        DocTestSuite('lovely.buildouthttp.buildouthttp',
                     setUp=setUp, tearDown=tearDown,
                     optionflags=doctest.NORMALIZE_WHITESPACE|doctest.ELLIPSIS,
                     ),
        )
    return unittest.TestSuite(uSuites)



bootstrap.py
# Open Source Software licensed under the ZPL license
# Copyright (c) 2006, Zope Foundation and Contributors.

##############################################################################
#
# Copyright (c) 2006 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Bootstrap a buildout-based project

Simply run this script in a directory containing a buildout.cfg.
The script accepts buildout command-line options, so you can
use the -c option to specify an alternate configuration file.
"""

import os, shutil, sys, tempfile, urllib, urllib2, subprocess
from optparse import OptionParser

if sys.platform == 'win32':
    def quote(c):
        if ' ' in c:
            return '"%s"' % c  # work around spawn lamosity on windows
        else:
            return c
else:
    quote = str

# See zc.buildout.easy_install._has_broken_dash_S for motivation and comments.
stdout, stderr = subprocess.Popen(
    [sys.executable, '-Sc',
     'try:\n'
     '    import ConfigParser\n'
     'except ImportError:\n'
     '    print 1\n'
     'else:\n'
     '    print 0\n'],
    stdout=subprocess.PIPE, stderr=subprocess.PIPE).communicate()
has_broken_dash_S = bool(int(stdout.strip()))

# In order to be more robust in the face of system Pythons, we want to
# run without site-packages loaded.  This is somewhat tricky, in
# particular because Python 2.6's distutils imports site, so starting
# with the -S flag is not sufficient.  However, we'll start with that:
if not has_broken_dash_S and 'site' in sys.modules:
    # We will restart with python -S.
    args = sys.argv[:]
    args[0:0] = [sys.executable, '-S']
    args = map(quote, args)
    os.execv(sys.executable, args)
# Now we are running with -S.  We'll get the clean sys.path, import site
# because distutils will do it later, and then reset the path and clean
# out any namespace packages from site-packages that might have been
# loaded by .pth files.
clean_path = sys.path[:]
import site  # imported because of its side effects
sys.path[:] = clean_path
for k, v in sys.modules.items():
    if k in ('setuptools', 'pkg_resources') or (
        hasattr(v, '__path__') and
        len(v.__path__) == 1 and
        not os.path.exists(os.path.join(v.__path__[0], '__init__.py'))):
        # This is a namespace package.  Remove it.
        sys.modules.pop(k)

is_jython = sys.platform.startswith('java')

setuptools_source = 'http://peak.telecommunity.com/dist/ez_setup.py'
distribute_source = 'http://python-distribute.org/distribute_setup.py'


# parsing arguments
def normalize_to_url(option, opt_str, value, parser):
    if value:
        if '://' not in value:  # It doesn't smell like a URL.
            value = 'file://%s' % (
                urllib.pathname2url(
                    os.path.abspath(os.path.expanduser(value))),)
        if opt_str == '--download-base' and not value.endswith('/'):
            # Download base needs a trailing slash to make the world happy.
            value += '/'
    else:
        value = None
    name = opt_str[2:].replace('-', '_')
    setattr(parser.values, name, value)

usage = '''\
[DESIRED PYTHON FOR BUILDOUT] bootstrap.py [options]

Bootstraps a buildout-based project.

Simply run this script in a directory containing a buildout.cfg, using the
Python that you want bin/buildout to use.

Note that by using --setup-source and --download-base to point to
local resources, you can keep this script from going over the network.
'''

parser = OptionParser(usage=usage)
parser.add_option("-v", "--version", dest="version",
                          help="use a specific zc.buildout version")
parser.add_option("-d", "--distribute",
                   action="store_true", dest="use_distribute", default=False,
                   help="Use Distribute rather than Setuptools.")
parser.add_option("--setup-source", action="callback", dest="setup_source",
                  callback=normalize_to_url, nargs=1, type="string",
                  help=("Specify a URL or file location for the setup file. "
                        "If you use Setuptools, this will default to " +
                        setuptools_source + "; if you use Distribute, this "
                        "will default to " + distribute_source + "."))
parser.add_option("--download-base", action="callback", dest="download_base",
                  callback=normalize_to_url, nargs=1, type="string",
                  help=("Specify a URL or directory for downloading "
                        "zc.buildout and either Setuptools or Distribute. "
                        "Defaults to PyPI."))
parser.add_option("--eggs",
                  help=("Specify a directory for storing eggs.  Defaults to "
                        "a temporary directory that is deleted when the "
                        "bootstrap script completes."))
parser.add_option("-t", "--accept-buildout-test-releases",
                  dest='accept_buildout_test_releases',
                  action="store_true", default=False,
                  help=("Normally, if you do not specify a --version, the "
                        "bootstrap script and buildout gets the newest "
                        "*final* versions of zc.buildout and its recipes and "
                        "extensions for you.  If you use this flag, "
                        "bootstrap and buildout will get the newest releases "
                        "even if they are alphas or betas."))
parser.add_option("-c", None, action="store", dest="config_file",
                   help=("Specify the path to the buildout configuration "
                         "file to be used."))

options, args = parser.parse_args()

if options.eggs:
    eggs_dir = os.path.abspath(os.path.expanduser(options.eggs))
else:
    eggs_dir = tempfile.mkdtemp()

if options.setup_source is None:
    if options.use_distribute:
        options.setup_source = distribute_source
    else:
        options.setup_source = setuptools_source

if options.accept_buildout_test_releases:
    args.insert(0, 'buildout:accept-buildout-test-releases=true')

try:
    import pkg_resources
    import setuptools  # A flag.  Sometimes pkg_resources is installed alone.
    if not hasattr(pkg_resources, '_distribute'):
        raise ImportError
except ImportError:
    ez_code = urllib2.urlopen(
        options.setup_source).read().replace('\r\n', '\n')
    ez = {}
    exec ez_code in ez
    setup_args = dict(to_dir=eggs_dir, download_delay=0)
    if options.download_base:
        setup_args['download_base'] = options.download_base
    if options.use_distribute:
        setup_args['no_fake'] = True
        if sys.version_info[:2] == (2, 4):
            setup_args['version'] = '0.6.32'
    ez['use_setuptools'](**setup_args)
    if 'pkg_resources' in sys.modules:
        reload(sys.modules['pkg_resources'])
    import pkg_resources
    # This does not (always?) update the default working set.  We will
    # do it.
    for path in sys.path:
        if path not in pkg_resources.working_set.entries:
            pkg_resources.working_set.add_entry(path)

cmd = [quote(sys.executable),
       '-c',
       quote('from setuptools.command.easy_install import main; main()'),
       '-mqNxd',
       quote(eggs_dir)]

if not has_broken_dash_S:
    cmd.insert(1, '-S')

find_links = options.download_base
if not find_links:
    find_links = os.environ.get('bootstrap-testing-find-links')
if not find_links and options.accept_buildout_test_releases:
    find_links = 'http://downloads.buildout.org/'
if find_links:
    cmd.extend(['-f', quote(find_links)])

if options.use_distribute:
    setup_requirement = 'distribute'
else:
    setup_requirement = 'setuptools'
ws = pkg_resources.working_set
setup_requirement_path = ws.find(
    pkg_resources.Requirement.parse(setup_requirement)).location
env = dict(
    os.environ,
    PYTHONPATH=setup_requirement_path)

requirement = 'zc.buildout'
version = options.version
if version is None and not options.accept_buildout_test_releases:
    # Figure out the most recent final version of zc.buildout.
    import setuptools.package_index
    _final_parts = '*final-', '*final'

    def _final_version(parsed_version):
        for part in parsed_version:
            if (part[:1] == '*') and (part not in _final_parts):
                return False
        return True
    index = setuptools.package_index.PackageIndex(
        search_path=[setup_requirement_path])
    if find_links:
        index.add_find_links((find_links,))
    req = pkg_resources.Requirement.parse(requirement)
    if index.obtain(req) is not None:
        best = []
        bestv = None
        for dist in index[req.project_name]:
            distv = dist.parsed_version
            if distv >= pkg_resources.parse_version('2dev'):
                continue
            if _final_version(distv):
                if bestv is None or distv > bestv:
                    best = [dist]
                    bestv = distv
                elif distv == bestv:
                    best.append(dist)
        if best:
            best.sort()
            version = best[-1].version

if version:
    requirement += '=='+version
else:
    requirement += '<2dev'

cmd.append(requirement)

if is_jython:
    import subprocess
    exitcode = subprocess.Popen(cmd, env=env).wait()
else:  # Windows prefers this, apparently; otherwise we would prefer subprocess
    exitcode = os.spawnle(*([os.P_WAIT, sys.executable] + cmd + [env]))
if exitcode != 0:
    sys.stdout.flush()
    sys.stderr.flush()
    print ("An error occurred when trying to install zc.buildout. "
           "Look above this message for any errors that "
           "were output by easy_install.")
    sys.exit(exitcode)

ws.add_entry(eggs_dir)
ws.require(requirement)
import zc.buildout.buildout

# If there isn't already a command in the args, add bootstrap
if not [a for a in args if '=' not in a]:
    args.append('bootstrap')


# if -c was provided, we push it back into args for buildout's main function
if options.config_file is not None:
    args[0:0] = ['-c', options.config_file]

zc.buildout.buildout.main(args)
if not options.eggs:  # clean up temporary egg directory
    shutil.rmtree(eggs_dir)


src/lovely/buildouthttp/buildouthttp.py
##############################################################################
#
# Copyright (c) 2007 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""
$Id: buildouthttp.py 116763 2010-09-23 12:16:01Z adamg $
"""
import pkg_resources
from six.moves.urllib import request as urllib_request
from six.moves.urllib import error as urllib_error
from six.moves.urllib import parse as urllib_parse
import sys

if sys.version_info[0] == 2:
    import urllib2
    url_opener = urllib2._opener
else:
    import urllib.request
    url_opener = urllib.request._opener

import os
import csv
import logging
import subprocess
from io import StringIO
from zc.buildout import download

log = logging.getLogger('lovely.buildouthttp')
original_build_opener = urllib_request.build_opener


def get_zc_buildout_version():
    """Determine the version of zc.buildout

    >>> len(get_zc_buildout_version()) > 0
    True

    >>> import re
    >>> re.match("\d+\.\d+\.\d+", get_zc_buildout_version()) is not None
    True
    """
    return pkg_resources.working_set.find(
        pkg_resources.Requirement.parse('zc.buildout')
    ).version


def get_github_credentials():

    """returns the credentials for the local git installation by using
    git config"""

    p = subprocess.Popen("git config github.accesstoken",
                         shell=True,
                         stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE)
    rc = p.wait()
    if rc:
        # failure to get config, so return silently
        return None
    token = p.stdout.readline().strip()
    if token:
        log.debug("Found github accesstoken %r", token)
        return token


def isPrivate(urlPath, repos):
    """Check if a URL is a private URL

    >>> isPrivate('/downloads/me/', None)
    True
    >>> isPrivate('/downloads/me/', [])
    False
    >>> isPrivate('/downloads/me/', ['me'])
    True
    >>> isPrivate('/repos/me/', ['me'])
    True
    >>> isPrivate('/downloads/me/', ['unknown'])
    False
    >>> isPrivate('/somewhere/me/', ['me'])
    False
    """
    # If provided check whitelist of Github repos in the format
    # "<userororg>/<repo>", for either API v3 or static downloads
    if repos is None:
        # for backward compatibility all repos are private if there is no
        # "github-repos" entry in the buildout section.
        return True
    # now check the whitelist
    for repo in repos:
        if not repo:
            # ignore empty entries
            continue
        api_repo = "/repos/%s/" % (repo,)
        dl_repo = "/downloads/%s/" % (repo,)
        if api_repo in urlPath or dl_repo in urlPath:
            return True
    return False


class GithubHandler(urllib_request.BaseHandler):
    """This handler creates a post request with login and token

    see http://github.com/blog/170-token-authentication for details

    With a none github url the resulting request is unchanged::

    >>> req = urllib_request.Request('http://example.com/downloads/me/')
    >>> handler = GithubHandler('--mytoken--')
    >>> res = handler.https_request(req)
    >>> res.get_full_url()
    'http://example.com/downloads/me/'
    >>> res is req
    True

    With a github url we get the access token::

    >>> req = urllib_request.Request('https://github.com/downloads/me/')
    >>> res = handler.https_request(req)
    >>> res.get_full_url()
    'https://github.com/downloads/me/?access_token=--mytoken--'

    If we provide an empty whitelist we get all github requests without a
    token::

    >>> handler = GithubHandler('--mytoken--', [])
    >>> req = urllib_request.Request('https://github.com/downloads/me/')
    >>> res = handler.https_request(req)
    >>> res.get_full_url()
    'https://github.com/downloads/me/'

    If the repository is in the whitelist is receives a token::

    >>> handler = GithubHandler('--mytoken--', ['me'])
    >>> req = urllib_request.Request('https://github.com/downloads/me/')
    >>> res = handler.https_request(req)
    >>> res.get_full_url()
    'https://github.com/downloads/me/?access_token=--mytoken--'

    >>> req = urllib_request.Request(
    ...           'https://github.com/downloads/me/?a=1&b=2')
    >>> res = handler.https_request(req)
    >>> res.get_full_url()
    'https://github.com/downloads/me/?a=1&b=2&access_token=--mytoken--'

    If no timeout is set in the original request, the timeout is set to 60::

    >>> hasattr(req, 'timeout')
    False
    >>> res.timeout
    60


    The timeout from the original request is preseverd in the result::

    >>> req = urllib_request.Request('https://github.com/downloads/me/?a=1&b=2')
    >>> req.timeout = 42
    >>> res = handler.https_request(req)
    >>> res.timeout
    42

    """

    def __init__(self, token, repos=None):
        self._token = token
        self._repos = repos

    def https_request(self, req):
        get_host = lambda: (req.get_host() if hasattr(req, "get_host") else req.host)
        if req.get_method() == 'GET' and get_host().endswith('github.com'):
            url = req.get_full_url()
            scheme, netloc, path, params, query, fragment = \
                                        urllib_parse.urlparse(url)
            if isPrivate(path, self._repos):
                log.debug("Found private github url %r", (url,))
                token = urllib_parse.urlencode(dict(access_token=self._token))
                query = '&'.join([p for p in (query, token) if p])
                new_url = urllib_parse.urlunparse((scheme, netloc, path, params,
                                               query, fragment))
                timeout = getattr(req, 'timeout', 60)
                old_req = req
                req = urllib_request.Request(new_url)
                req.timeout = timeout
                # Re-add user-agent, as the GitHub API requires this for auth
                req.add_header('user-agent', old_req.get_header('user-agent',
                                                            'Python-urllib2'))
            else:
                log.debug("Github url %r blocked by buildout.github-repos" %
                          (url,))
                log.debug(self._repos)
        return req


class CredHandler(urllib_request.HTTPBasicAuthHandler):

    """This handler adds basic auth credentials to the request upon a 401

    >>> auth_handler = CredHandler()
    >>> auth_handler.add_password('myrealm', 'http://example.com',
    ...                           'user', 'password')

    >>> from six import StringIO
    >>> fp = StringIO('The error body')
    >>> req = urllib_request.Request('http://example.com')
    >>> auth_handler.http_error_401(req, fp, 401, 'msg', {})
    """

    def http_error_401(self, req, fp, code, msg, headers):
        try:
            #python 2.6 introduces this attribute and fails the request on 5
            #but since we're process global this fails on the 5th download
            #there's no reset_retry_count(), so clear it here:
            self.retried = 0
        except AttributeError:
            pass

        log.debug('getting url: %r' % req.get_full_url())
        try:
            res = urllib_request.HTTPBasicAuthHandler.http_error_401(
                self, req, fp, code, msg, headers)
        except urllib_error.HTTPError as err:
            log.error('failed to get url: %r %r', req.get_full_url(), err.code)
            raise
        except Exception as err:
            log.error('failed to get url: %r %s', req.get_full_url(), str(err))
            raise
        else:
            if res is None:
                log.error('failed to get url: %r, check your realm',
                          req.get_full_url())
            elif res.code >= 400:
                log.error('failed to get url: %r %r', res.url, res.code)
            else:
                log.debug('got url: %r %r', res.url, res.code)
            return res


def unload(buildout=None):
    urllib_request.build_opener = original_build_opener
    urllib_request.install_opener(urllib_request.build_opener())


def install(buildout=None, pwd_path=None):
    pwdsf = StringIO()
    combined_creds = []
    github_creds = None
    creds = []
    local_pwd_path = ''
    github_repos = None
    if buildout is not None:
        local_pwd_path = os.path.join(
            buildout['buildout']['directory'],
            '.httpauth')
        if 'github-repos' in buildout['buildout']:
            github_repos = buildout['buildout']['github-repos'].split('\n')
    system_pwd_path = os.path.join(
        os.path.expanduser('~'),
        '.buildout',
        '.httpauth')

    def combine_cred_file(file_path, combined_creds):
        if file_path is None or not os.path.exists(file_path):
            return
        cred_file = open(file_path)
        combined_creds += [l.strip()
                            for l in cred_file.readlines() if l.strip()]
        cred_file.close()
    # combine all the possible .httpauth files together
    combine_cred_file(pwd_path, combined_creds)
    combine_cred_file(local_pwd_path, combined_creds)
    combine_cred_file(system_pwd_path, combined_creds)
    pwdsf_len = pwdsf.write(u"\n".join(combined_creds))
    pwdsf.seek(0)
    if not pwdsf_len:
        pwdsf = None
        log.warn('Could not load authentication information')
    try:
        auth_handler = CredHandler()
        github_creds = get_github_credentials()
        new_handlers = []
        if github_creds:
            new_handlers.append(GithubHandler(github_creds, github_repos))
        if pwdsf:
            for l, row in enumerate(csv.reader(pwdsf)):
                if len(row) == 3:
                    realm, uris, user = (el.strip() for el in row)
                    password = prompt_passwd(realm, user)
                elif len(row) == 4:
                    realm, uris, user, password = (el.strip() for el in row)
                else:
                    raise RuntimeError(
                        "Authentication file cannot be parsed %s:%s" % (
                            pwd_path, l + 1))
                creds.append((realm, uris, user, password))
                log.debug('Added credentials %r, %r' % (realm, uris))
                auth_handler.add_password(realm, uris, user, password)
        if creds:
            new_handlers.append(auth_handler)
        if creds or github_creds:
            if get_zc_buildout_version() >= '2.0.0':
                download.urllib._urlopener = URLOpener(creds, github_creds, github_repos)
            else:
                download.url_opener = URLOpener(creds, github_creds, github_repos)
        if new_handlers:
            if url_opener is not None:
                handlers = url_opener.handlers[:]
                handlers[:0] = new_handlers
            else:
                handlers = new_handlers
            urllib_request.build_opener = lambda *a: original_build_opener(*handlers)
            urllib_request.install_opener(urllib_request.build_opener())
    finally:
        if pwdsf:
            pwdsf.close()


def prompt_passwd(realm, user):
    from getpass import getpass
    password = getpass('>>> Password for {} - {}: '.format(realm, user))
    return password


class URLOpener(download.URLOpener):

    def __init__(self, creds, github_creds, github_repos):
        self.creds = {}
        self.github_creds = github_creds
        self.github_repos = github_repos
        for realm, uris, user, password in creds:
            parts = urllib_parse.urlparse(uris)
            self.creds[(parts[1], realm)] = (user, password)
        download.URLOpener.__init__(self)

    def retrieve(self, url, filename=None, reporthook=None, data=None):
        if self.github_creds and not data:
            scheme, netloc, path, params, query, fragment = urllib_parse.urlparse(
                url)
            if (scheme == 'https'
                and netloc.endswith('github.com')
                and isPrivate(url, self.github_repos)
               ):
                log.debug("Appending github credentials to url %r", url)
                token = self.github_creds
                cred = urllib_parse.urlencode(dict(access_token=token))
                query = '&'.join((query, cred))
            url = urllib_parse.urlunparse((scheme, netloc, path, params,
                                       query, fragment))
        return download.URLOpener.retrieve(self, url, filename,
                                           reporthook, data)

    def prompt_user_passwd(self, host, realm):
        creds = self.creds.get((host, realm))
        if creds:
            return creds
        return download.URLOpener.prompt_user_passwd(self, host, realm)
